src/planning/include/planning/Navfn.hpp

```diff
@@ -0,0 +1,267 @@
+*  COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+*  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+*  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+*  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+*  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+*  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+*  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+*  POSSIBILITY OF SUCH DAMAGE.
+*********************************************************************/
+
+//
+// Navigation function computation
+// Uses Dijkstra's method
+// Modified for Euclidean-distance computation
+//
+
+#ifndef _NAVFN_H
+#define _NAVFN_H
+
+#include <math.h>
+#include <stdint.h>
+#include <string.h>
+#include <stdio.h>
+
+// cost defs
+#define COST_UNKNOWN_ROS 255		// 255 is unknown cost
+#define COST_OBS 254		// 254 for forbidden regions
+#define COST_OBS_ROS 253	// ROS values of 253 are obstacles
+
+// navfn cost values are set to
+// COST_NEUTRAL + COST_FACTOR * costmap_cost_value.
+// Incoming costmap cost values are in the range 0 to 252.
+// With COST_NEUTRAL of 50, the COST_FACTOR needs to be about 0.8 to
+// ensure the input values are spread evenly over the output range, 50
+// to 253.  If COST_FACTOR is higher, cost values will have a plateau
+// around obstacles and the planner will then treat (for example) the
+// whole width of a narrow hallway as equally undesirable and thus
+// will not plan paths down the center.
+
+#define COST_NEUTRAL 50		// Set this to "open space" value
+#define COST_FACTOR 0.8		// Used for translating costs in NavFn::setCostmap()
+
+// Define the cost type in the case that it is not set. However, this allows
+// clients to modify it without changing the file. Arguably, it is better to require it to
+// be defined by a user explicitly
+#ifndef COSTTYPE
+#define COSTTYPE unsigned char	// Whatever is used...
+#endif
+
+// potential defs
+#define POT_HIGH 1.0e10		// unassigned cell potential
+
+// priority buffers
+#define PRIORITYBUFSIZE 10000
+
+
+// namespace navfn {
+  /**
+    Navigation function call.
+    \param costmap Cost map array, of type COSTTYPE; origin is upper left
+NOTE: will be modified to have a border of obstacle costs
+\param nx Width of map in cells
+\param ny Height of map in cells
+\param goal X,Y position of goal cell
+\param start X,Y position of start cell
+
+Returns length of plan if found, and fills an array with x,y interpolated
+positions at about 1/2 cell resolution; else returns 0.
+
+*/
+
+  int create_nav_plan_astar(const COSTTYPE *costmap, int nx, int ny,
+      int* goal, int* start,
+      float *plan, int nplan);
+
+
+
+  /**
+   * @class NavFn
+   * @brief Navigation function class. Holds buffers for costmap, navfn map. Maps are pixel-based. Origin is upper left, x is right, y is down. 
+   */
+  class NavFn
```

Comment:
We should refactor this class. It's not clear to me what it is. We can probably extract out the various ideas into different (reusable) classes. Doesn't have to be done immediately though.